system/lib/wasmfs/wasmfs.cpp

```diff
@@ -56,16 +56,52 @@ WasmFS::WasmFS() : rootDirectory(initRootDirectory()), cwd(rootDirectory) {
 // (Note that this means we can't find leaks inside WasmFS code itself, but that
 // seems fundamentally impossible for the above reasons, unless we made LSan log
 // its findings in a way that does not depend on normal file I/O.)
-__attribute__((weak)) extern "C" void __lsan_do_leak_check(void) {
-}
+__attribute__((weak)) extern "C" void __lsan_do_leak_check(void) {}
 
 extern "C" void wasmfs_flush(void) {
   // Flush musl libc streams.
   fflush(0);
 
-  // Flush our own streams. TODO: flush all backends.
+  // Flush our own streams.
   (void)SpecialFiles::getStdout()->locked().flush();
   (void)SpecialFiles::getStderr()->locked().flush();
+
+  std::vector<std::shared_ptr<Directory>> toFlush;
+  toFlush.push_back(wasmFS.getRootDirectory());
+
+  while (!toFlush.empty()) {
+    auto dir = toFlush.back();
+    toFlush.pop_back();
+
+    auto lockedDir = dir->locked();
+    Directory::MaybeEntries entries = lockedDir.getEntries();
+#ifndef NDEBUG
+    if (int err = entries.getError()) {
```

Comment:
I think we should check for an error even in release mode (not just in debug), but just not show a verbose text error. That is, if an error happened, we should return out of this function. Otherwise we'd continue onwards and access the entries when it's not valid to do so, which could be very bad.

Reply:
Ah, I see the potential issue. But would it make more sense for the function to directly return or just continue to the next item in the stack? I think I could see a situation where one directory errors but the other directories can still be flushed.

Reply:
Ah, right, it's ok to continue with what is already on the stack, as those can't be nested under the erroring directory. Good point.